### DISCUSSION TITLE
Alienware on the way.

### DISCUSSION THREAD

**Discussion by T-Troll** [DadouLapous/DellG5SE-Fan-Linux#6] [Q&A] [answered]
Thanks for you tool, it's inspire me to do the same for Alienware notebooks (now there is no way to control fans except AWCC).

I check RWE, and configure out register map is completely different.
I found temps, fan modes (steps) and RPMs, even fan boost register, but still can't locate unlock ones.

Any hints, please? What should be into it upon restart?

**Comment by DavidLapous** [comment]
Thanks for your interest. The "unlock" registry offsets are indeed the hardest to find as they do not change over time (unless there is a program in windows that can. For msi's laptop for instance, there is a manual fan mode in msi dragon center)... 
There are some way to find them though. (Well documented in this (dead ?) [project](https://github.com/hirschmann/nbfc/wiki/How-to-create-a-NBFC-config) ) Notice that finding this "unlock" mode *may* not be necessary, on my laptop for instance, most of the time, right fan follows the asked fan speed without beeing in "unlocked mode". The methods I know :
1. This work has already be done on this laptop, and you only need to copy pasta the offsets. For instance [nbfc](https://github.com/hirschmann/nbfc/tree/master/Configs) or [isw](https://github.com/YoyPa/isw) (for msi's laptop) have some examples of working configurations. Even if your laptop isn't in there, it gives you ideas on where this address tends to be (on similar laptops). For instance, on my laptop, this manual trigger is close to fan target speed, as mentioned in [this file](https://github.com/DavidLapous/DellG5SE-Fan-Linux/blob/main/EC_addresses). Disabled value is `0x04` and enabled is `0x14`. "Similar" can be very different; for instance, the Xiaomi mi book, from  nbfc's config list,  share most of the offsets with the dell g5 se (sadly I noticed that after finding these values myself).
2. If you haven't success, you can also search for your EC's datasheet on the internet (if it exists) -- (look at the nbfc's wiki) (I had no success here)
3. Or decompile  your laptop's dsdt. (with `iasl` on linux, in the `acpica`' package), again, this is detailed in the nbfc's wiki. More resources [here](https://wiki.archlinux.org/title/DSDT) (on linux). (I didn't have much success here, but you can find useful stuff here)
4. If these methods didn't give any result, you'll have to find it manually. I don't have any magic tricks here, that's a trial and error thing. You have to find possible candidates (this offset wont move over time, whatever you're doing to the laptop (ie. unplug, cpu stress, numpad on/off, change backlight/rgb keyboard, boost fan...) , should be close to fan related variables, hopefully nonzero, ...), try move them and see how they behave (revert ASAP if that's not the good one), and try changing fan speeds. That may take a long time.

Fortunately, same vendors tends to use similar ECs. 

Good luck with your laptop ! Replacing AWCC is a really great idea :).

**Reply by DavidLapous** [reply]
An idea of how that's presented on my laptop (only the fan related part)
![image](https://user-images.githubusercontent.com/15857585/127632348-788fef77-254c-4f30-bd78-4ee71b0df752.png)

**Reply by T-Troll** [reply]
Thanks a lot!
You make me busy for a weekend, indeed!

I don't target for exactly my notebooks, but will try to make something universal, for any Dell G/Alienware. Like i do for lights. 
BTW - feel free to port my SDK to Linux if you want light control (you rig is supported) - it's simple USBHID reports/features.

I'll share the results if any success. For now, i can only enable boost mode - all fans 110%. Maybe it will also work for you gear.

**Reply by DavidLapous** [reply]
Nice project :)! sadly my laptop wasn't delivered with RGB / lights, so I won't be able to do that ^^. If you have the offsets of other laptops though, I'll add them to my repository to support more (similar) laptops. "Similar" meaning : if their fancontrol is completely different (ie. significantly more work than changing offsets) I don't think I'll take the time to make another fan curve logic.

**Reply by T-Troll** [reply]
> An idea of how that's presented on my laptop (only the fan related part)

Here is for you as an exchange:
[EC6662_quiet.rw.zip](https://github.com/DavidLapous/DellG5SE-Fan-Linux/files/6906537/EC6662_quiet.rw.zip)
It's Alienware m15R1 in quiet mode.

What do i know now:
0xEC, 0xED - target fan speed (0..6).
0xEE - CPU temp
0xFD - GPU temp
0xe8/e9, 0xea/eb - fan's RPM (see LM78A specs - it's in tachometer count units)
0xE7 - fan boost. Any value here provide fans run up to 110%. Still don't know how to disable it, instead of reboot. :-)

**Reply by DavidLapous** [reply]
BTW there is (to my knowledge) no fan boost offset on the dellg5se on the EC, I simply put the fan speeds to the boost speed on my script. So no overclocking of cpu/dgpu with my `-b` option.

**Reply by T-Troll** [reply]
Ok, seems like you will be busy a bit too. %-)
Try to look around - as you can see (and i check before) G5 have the same chip as m15/m17 (any revision), just different ACPI mappings.

**Reply by T-Troll** [reply]
Update.
So, i dig deep into the DSDT... Alienware guys are strange.
First, they have 3 EC blocks. One - accessible by common EC port/data register is a mix - first half (00-80) is, in fact, mapped EC from Nvidia GPU. Second part is "common" EC, but... it's mapped from memory. And other one - memory-mapped only.

The third part have all control registers and interfaces, so you need access physical RAM instead of EC to set something up. Not so easy in Windows, by the way.

I RE all protocol and can control fans and do many other tricks (like control CPU power limit - yes, AW have this possibility). So i develop a proof tool now, polish and can share protocol - if you want to implement it.

PS: I not miss any EC offsets - except what i named "boost". It's real name - TERR (Thermal Error). It enables emergency mode, in fact, that's why it can't be disabled. ^_^

**Reply by DavidLapous** [reply]
Nice work !! For the cpu power limit and temps, as the dellg5se's uses amd, there is the fantastic FlyGoat's [ryzenadj](https://github.com/FlyGoat/RyzenAdj) which handle that perfectly. The GPU part (amdgpu) has an opensource driver on linux, so it's also a nonissue on this laptop :). I'll still take a look on your work to see if there is some stuff that can be useful on this laptop (or another, if I expand this tool to other laptops) !

**Reply by T-Troll** [reply]
Thank you.
But here is the dragons. As you can see, i have a community of my project users. I've ask them to share info for me... Well, Alienware is a zoo. Nothing similar at low level (i know it then work for lights.... 5 different API approaches). The only common think i catch is AWCC ACPI interface. So.... I going this way. Direct ACPI calls. It's more or less universal (between Alienware gear), at least.
Please take my respect inspiring me to dig into this and provide critical info to understand. Feel free to port any of my tools (it will be fan tool soon as well) - i'm Windows guy, not Linux (instead i play embedded Linux at my main duty).
It was great!

**Reply by DavidLapous** [reply]
You're right, when working, acpi calls should be the best option. On my laptop though awcc's fan control isn't even working, I hope that's not such a mess on other dell laptops. This is my first dell laptop, so I won't be able to help you much more... 
Specifically to amd's cpu/gpu :
- To my knowledge, ryzenadj can handle every laptop having a ryzen cpu (works also on windows).  
- For the dgpu part `MorePowerTool` should be working, but I didn't dig into it. I don't know if it provides a library to rely on it.

**Reply by T-Troll** [reply]
Here we go:
https://github.com/T-Troll/alienfan-tools
Now support only m15/17R1 series, and it's THE ONLY tool can support it AFAIK (both DFC, SpeedFan, HWInfo and a bunch of other "common" tools can't).

So thank you for the hints, feel free to use my ACPI call codes as well.

**Comment by T-Troll** [comment, marked as answer]
Thank you for help, I will open a topic about GUI to share my approach.
Great job!
